From 0f230ad0e42ae3e2e76b32dca86e7007c81eef4d Mon Sep 17 00:00:00 2001
From: Jernej Skrabec <jernej.skrabec@siol.net>
Date: Tue, 5 Jan 2021 17:42:35 +0100
Subject: [PATCH] cir: Replace precalculated numbers with macros

RC6 duration table contains precalculated numbers which are not
explained. Replace them by macros which are more readable. Compiler will
lower them to constants anyway.

Signed-off-by: Jernej Skrabec <jernej.skrabec@siol.net>
---
 drivers/cir/sunxi-cir.c | 25 ++++++++++++++++---------
 1 file changed, 16 insertions(+), 9 deletions(-)

diff --git a/drivers/cir/sunxi-cir.c b/drivers/cir/sunxi-cir.c
index 2f462858..089dbcc1 100644
--- a/drivers/cir/sunxi-cir.c
+++ b/drivers/cir/sunxi-cir.c
@@ -21,6 +21,14 @@
 #define CIR_RXSTA  0x30
 #define CIR_RXCFG  0x34
 
+#define CIR_CLK_RATE 32768UL
+
+/* RC6 time unit is 16 periods @ 36 kHz, ~444 us */
+#define RC6_TIME_UNIT 444UL
+
+/* convert specified number of time units to number of clock cycles */
+#define UNITS_TO_CLKS(num) (((num) * CIR_CLK_RATE * RC6_TIME_UNIT) / 1000000UL)
+
 struct sunxi_cir_state {
 	struct device_state ds;
 	struct rc6_ctx      rc6_ctx;
@@ -29,16 +37,15 @@ struct sunxi_cir_state {
 	uint32_t            ctl_stash;
 };
 
-/* These durations are based on a 32768 Hz sample clock. */
 static const int8_t sunxi_cir_rc6_durations[RC6_STATES] = {
-	[RC6_IDLE]      = 6 * 14,
-	[RC6_LEADER_S]  = 2 * 14,
-	[RC6_HEADER_P]  = 1 * 14,
-	[RC6_HEADER_N]  = 1 * 14,
-	[RC6_TRAILER_P] = 2 * 14,
-	[RC6_TRAILER_N] = 2 * 14,
-	[RC6_DATA_P]    = 1 * 14,
-	[RC6_DATA_N]    = 1 * 14,
+	[RC6_IDLE]      = UNITS_TO_CLKS(6),
+	[RC6_LEADER_S]  = UNITS_TO_CLKS(2),
+	[RC6_HEADER_P]  = UNITS_TO_CLKS(1),
+	[RC6_HEADER_N]  = UNITS_TO_CLKS(1),
+	[RC6_TRAILER_P] = UNITS_TO_CLKS(2),
+	[RC6_TRAILER_N] = UNITS_TO_CLKS(2),
+	[RC6_DATA_P]    = UNITS_TO_CLKS(1),
+	[RC6_DATA_N]    = UNITS_TO_CLKS(1),
 };
 
 static inline const struct sunxi_cir *
From 00dac6bd96ea1d754d0104f4800a9fead3265eaa Mon Sep 17 00:00:00 2001
From: Jernej Skrabec <jernej.skrabec@siol.net>
Date: Wed, 6 Jan 2021 00:50:51 +0100
Subject: [PATCH] cir: Expand durations to 16-bit

Currently all durations fit in signed 8-bit variable. However, this will
change in following commit with higher clock rate.

Make durations 16-bit.

Signed-off-by: Jernej Skrabec <jernej.skrabec@siol.net>
---
 drivers/cir/rc6.h       | 12 ++++++------
 drivers/cir/sunxi-cir.c |  2 +-
 2 files changed, 7 insertions(+), 7 deletions(-)

diff --git a/drivers/cir/rc6.h b/drivers/cir/rc6.h
index e49a391d..7c16c701 100644
--- a/drivers/cir/rc6.h
+++ b/drivers/cir/rc6.h
@@ -21,12 +21,12 @@ enum {
 };
 
 struct rc6_ctx {
-	const int8_t *durations;
-	uint32_t      buffer;
-	uint8_t       bits;
-	uint8_t       state;
-	uint8_t       pulse;
-	int8_t        width;
+	const int16_t *durations;
+	uint32_t       buffer;
+	uint8_t        bits;
+	uint8_t        state;
+	uint8_t        pulse;
+	int8_t         width;
 };
 
 /**
diff --git a/drivers/cir/sunxi-cir.c b/drivers/cir/sunxi-cir.c
index 089dbcc1..6fdb80c4 100644
--- a/drivers/cir/sunxi-cir.c
+++ b/drivers/cir/sunxi-cir.c
@@ -37,7 +37,7 @@ struct sunxi_cir_state {
 	uint32_t            ctl_stash;
 };
 
-static const int8_t sunxi_cir_rc6_durations[RC6_STATES] = {
+static const int16_t sunxi_cir_rc6_durations[RC6_STATES] = {
 	[RC6_IDLE]      = UNITS_TO_CLKS(6),
 	[RC6_LEADER_S]  = UNITS_TO_CLKS(2),
 	[RC6_HEADER_P]  = UNITS_TO_CLKS(1),
From 54f0052f8885bd993d77052ddc237cafc29ee100 Mon Sep 17 00:00:00 2001
From: Jernej Skrabec <jernej.skrabec@siol.net>
Date: Tue, 5 Jan 2021 17:49:56 +0100
Subject: [PATCH] cir: Add support for OSC24M parent clock

Some boards come without external 32768 Hz oscillator. In such cases RTC
uses internal RC oscillator for generating osc32k clock. However, it is
not accurate enough for CIR needs.

Add an option to use OSC24M as a parent instead. Note that this will
increase power consumption in suspended/shutdown state.

Signed-off-by: Jernej Skrabec <jernej.skrabec@siol.net>
---
 drivers/cir/Kconfig     | 13 +++++++++++++
 drivers/cir/sunxi-cir.c | 16 +++++++++++++---
 2 files changed, 26 insertions(+), 3 deletions(-)

diff --git a/drivers/cir/Kconfig b/drivers/cir/Kconfig
index b440e81f..5fed465c 100644
--- a/drivers/cir/Kconfig
+++ b/drivers/cir/Kconfig
@@ -20,6 +20,19 @@ config CIR_WAKE_CODE
 
 		The default value will work with an RC6 MCE remote controller.
 
+config CIR_USE_OSC24M
+	bool "Use OSC24M as parent clock"
+	select NEED_OSC24M
+	default n
+	help
+		Sometimes boards come without external 32768 Hz oscillator and
+		thus RTC use internal RC oscilator as a source for generating
+		32768 Hz clock. However, such clock is not precise enough for
+		RC6 IR signal decoding purposes.
+
+		Set this option to y if your board doesn't have external 32768
+		Hz oscillator. Note that this increases power consumption.
+
 config R_CIR_RX_PIN
 	int
 	default  9 if PLATFORM_H6
diff --git a/drivers/cir/sunxi-cir.c b/drivers/cir/sunxi-cir.c
index 6fdb80c4..630b113e 100644
--- a/drivers/cir/sunxi-cir.c
+++ b/drivers/cir/sunxi-cir.c
@@ -21,7 +21,12 @@
 #define CIR_RXSTA  0x30
 #define CIR_RXCFG  0x34
 
+#if CONFIG(CIR_USE_OSC24M)
+/* parent clock is predivided by 192 */
+#define CIR_CLK_RATE 125000UL
+#else
 #define CIR_CLK_RATE 32768UL
+#endif
 
 /* RC6 time unit is 16 periods @ 36 kHz, ~444 us */
 #define RC6_TIME_UNIT 444UL
@@ -90,10 +95,15 @@ sunxi_cir_probe(const struct device *dev UNUSED)
 	state->rc6_ctx.durations = sunxi_cir_rc6_durations;
 
 	state->clk_stash = mmio_read_32(R_CIR_RX_CLK_REG);
-	mmio_write_32(R_CIR_RX_CLK_REG, 0x80000000);
-
 	state->cfg_stash = mmio_read_32(self->regs + CIR_RXCFG);
-	mmio_write_32(self->regs + CIR_RXCFG, 0x010f0310);
+
+	if (CONFIG(CIR_USE_OSC24M)) {
+		mmio_write_32(R_CIR_RX_CLK_REG, 0x81000002);
+		mmio_write_32(self->regs + CIR_RXCFG, 0x00001404);
+	} else {
+		mmio_write_32(R_CIR_RX_CLK_REG, 0x80000000);
+		mmio_write_32(self->regs + CIR_RXCFG, 0x010f0310);
+	}
 
 	state->ctl_stash = mmio_read_32(self->regs + CIR_RXCTL);
 	mmio_write_32(self->regs + CIR_RXCTL, 0x30);
From c1bb8e464960f4ef9cf45f6409aa69c6a99d818f Mon Sep 17 00:00:00 2001
From: Jernej Skrabec <jernej.skrabec@siol.net>
Date: Mon, 4 Jan 2021 21:31:28 +0100
Subject: [PATCH] configs: Add Tanix TX6 config

This is cheap TV box based on H6. Most notably it comes without PMIC and
32768 Hz external oscillator.

Signed-off-by: Jernej Skrabec <jernej.skrabec@siol.net>
---
 configs/tanix_tx6_defconfig | 3 +++
 1 file changed, 3 insertions(+)
 create mode 100644 configs/tanix_tx6_defconfig

diff --git a/configs/tanix_tx6_defconfig b/configs/tanix_tx6_defconfig
new file mode 100644
index 00000000..e7cc10a7
--- /dev/null
+++ b/configs/tanix_tx6_defconfig
@@ -0,0 +1,5 @@
+CONFIG_PLATFORM_H6=y
+CONFIG_CIR=y
+CONFIG_CIR_USE_OSC24M=y
+CONFIG_CIR_NEC=y
+CONFIG_CIR_WAKE_CODE=0x8051
